File: PARTICIPACIONES/part-4/Circulo2D.java
public class Circulo2D {
   
    private double x;
    private double y;
    private double radio;
    
   
    public Circulo2D() {
        this(0, 0, 1);
    }
    
    
    public Circulo2D(double x, double y, double radio) {
        this.x = x;
        this.y = y;
        this.radio = radio;
    }
    
    
    public double getX() {
        return x;
    }
    
    public double getY() {
        return y;
    }
    
    public double getRadio() {
        return radio;
    }
    
    
    public double getArea() {
        return Math.PI * radio * radio;
    }
    
    
    public double getPerimetro() {
        return 2 * Math.PI * radio;
    }
    
    
    public boolean contiene(double x, double y) {
        double distancia = Math.sqrt(Math.pow(x - this.x, 2) + Math.pow(y - this.y, 2));
        return distancia <= radio;
    }
    
    
    public boolean contiene(Circulo2D circulo) {
        double distanciaCentros = Math.sqrt(Math.pow(circulo.getX() - this.x, 2) + 
                                           Math.pow(circulo.getY() - this.y, 2));
        return distanciaCentros + circulo.getRadio() <= this.radio;
    }
    
    
    public boolean sobrepone(Circulo2D circulo) {
        double distanciaCentros = Math.sqrt(Math.pow(circulo.getX() - this.x, 2) + 
                                           Math.pow(circulo.getY() - this.y, 2));
        return distanciaCentros < (this.radio + circulo.getRadio()) && 
               !this.contiene(circulo) && 
               !circulo.contiene(this);
    }
    
    
    public static void main(String[] args) {
        
        Circulo2D c1 = new Circulo2D(2, 0, 1);
        
        System.out.println("Área de c1: " + c1.getArea());
        System.out.println("Perímetro de c1: " + c1.getPerimetro());
        System.out.println();
        
        
        System.out.println("c1.contiene(2.5, 0): " + c1.contiene(2.5, 0));
        
        
        Circulo2D circuloPequeño = new Circulo2D(2, 0, 0.5);
        System.out.println("c1.contiene(Circulo2D(2, 0, 0.5)): " + c1.contiene(circuloPequeño));
        
        
        Circulo2D circuloGrande = new Circulo2D(0, 0, 2);
        System.out.println("c1.sobrepone(Circulo2D(0, 0, 2)): " + c1.sobrepone(circuloGrande));
    }
}

File: PARTICIPACIONES/part-4/circulo2d.py
import math

class Circulo2D:
    
    def __init__(self, x=0, y=0, radio=1):
        self.__x = x
        self.__y = y
        self.__radio = radio
    
    
    def get_x(self):
        return self.__x
    
    def get_y(self):
        return self.__y
    
    def get_radio(self):
        return self.__radio
    
    
    def get_area(self):
        return math.pi * self.__radio * self.__radio
    
    
    def get_perimetro(self):
        return 2 * math.pi * self.__radio
    
    
    def contiene(self, x, y):
        distancia = math.sqrt((x - self.__x) ** 2 + (y - self.__y) ** 2)
        return distancia <= self.__radio
    
    
    def contiene_circulo(self, circulo):
        distancia_centros = math.sqrt((circulo.get_x() - self.__x) ** 2 + 
                                     (circulo.get_y() - self.__y) ** 2)
        return distancia_centros + circulo.get_radio() <= self.__radio
    
    
    def sobrepone(self, circulo):
        distancia_centros = math.sqrt((circulo.get_x() - self.__x) ** 2 + 
                                     (circulo.get_y() - self.__y) ** 2)
        return (distancia_centros < (self.__radio + circulo.get_radio()) and 
                not self.contiene_circulo(circulo) and 
                not circulo.contiene_circulo(self))


if __name__ == "__main__":
    c1 = Circulo2D(2, 0, 1)
    
    print("Área de c1: " + str(round(c1.get_area(), 2)))
    print("Perímetro de c1: " + str(round(c1.get_perimetro(), 2)))
    print()
    
    print("c1.contiene(2.5, 0): " + str(c1.contiene(2.5, 0)))
    
    circulo_pequeño = Circulo2D(2, 0, 0.5)
    print("c1.contiene(Circulo2D(2, 0, 0.5)): " + str(c1.contiene_circulo(circulo_pequeño)))
    
    circulo_grande = Circulo2D(0, 0, 2)
    print("c1.sobrepone(Circulo2D(0, 0, 2)): " + str(c1.sobrepone(circulo_grande)))